File: middleware/node/src/logger.ts
/**
 * In-memory call logger for the x402 middleware.
 *
 * Every successful paid call adds an entry to a ring buffer, which the
 * dashboard reads via /x402/stats. Lightweight, no DB required, bounded
 * memory (default 500 entries). For production with multiple backend
 * instances, swap this for Redis or a proper DB.
 */

export interface PaidCallLog {
  timestamp: number; // epoch ms
  endpoint: string;
  payer: string;
  amount: number; // USDC
  facilitator: string; // "openzeppelin" or "dry-run"
  durationMs: number;
}

export interface StatsSummary {
  totalCalls: number;
  totalRevenue: number;
  uniquePayers: number;
  byEndpoint: Array<{ endpoint: string; calls: number; revenue: number }>;
  recent: PaidCallLog[];
}

export class CallLogger {
  private buffer: PaidCallLog[] = [];
  private max: number;

  constructor(maxEntries = 500) {
    this.max = maxEntries;
  }

  log(entry: PaidCallLog): void {
    this.buffer.unshift(entry);
    if (this.buffer.length > this.max) {
      this.buffer.length = this.max;
    }
  }

  stats(limit = 50): StatsSummary {
    const totalCalls = this.buffer.length;
    const totalRevenue = this.buffer.reduce((s, e) => s + e.amount, 0);
    const uniquePayers = new Set(this.buffer.map((e) => e.payer)).size;

    const byEp = new Map<string, { calls: number; revenue: number }>();
    for (const e of this.buffer) {
      const cur = byEp.get(e.endpoint) ?? { calls: 0, revenue: 0 };
      cur.calls += 1;
      cur.revenue += e.amount;
      byEp.set(e.endpoint, cur);
    }
    const byEndpoint = Array.from(byEp.entries())
      .map(([endpoint, v]) => ({ endpoint, ...v }))
      .sort((a, b) => b.revenue - a.revenue);

    return {
      totalCalls,
      totalRevenue: Math.round(totalRevenue * 10000) / 10000,
      uniquePayers,
      byEndpoint,
      recent: this.buffer.slice(0, limit),
    };
  }
}

// Singleton — the x402 middleware and the dashboard both import this.
export const defaultLogger = new CallLogger();

File: middleware/python/x402_middleware/logger.py
"""In-memory call logger for the dashboard."""

from __future__ import annotations

import time
from collections import deque, defaultdict
from dataclasses import dataclass, asdict
from typing import Any


@dataclass
class PaidCallLog:
    timestamp: float  # unix seconds
    endpoint: str
    payer: str
    amount: float  # USDC
    facilitator: str
    duration_ms: int


class CallLogger:
    def __init__(self, max_entries: int = 500):
        self._buffer: deque[PaidCallLog] = deque(maxlen=max_entries)

    def log(self, entry: PaidCallLog) -> None:
        self._buffer.appendleft(entry)

    def stats(self, limit: int = 50) -> dict[str, Any]:
        total_calls = len(self._buffer)
        total_revenue = round(sum(e.amount for e in self._buffer), 4)
        unique_payers = len({e.payer for e in self._buffer})

        by_ep: dict[str, dict[str, float]] = defaultdict(lambda: {"calls": 0, "revenue": 0.0})
        for e in self._buffer:
            by_ep[e.endpoint]["calls"] += 1
            by_ep[e.endpoint]["revenue"] += e.amount

        by_endpoint = sorted(
            [{"endpoint": k, "calls": v["calls"], "revenue": round(v["revenue"], 4)} for k, v in by_ep.items()],
            key=lambda x: -x["revenue"],
        )

        recent = [
            {
                "timestamp": int(e.timestamp * 1000),  # JS ms to match Node dashboard
                "endpoint": e.endpoint,
                "payer": e.payer,
                "amount": e.amount,
                "facilitator": e.facilitator,
                "durationMs": e.duration_ms,
            }
            for e in list(self._buffer)[:limit]
        ]

        return {
            "totalCalls": total_calls,
            "totalRevenue": total_revenue,
            "uniquePayers": unique_payers,
            "byEndpoint": by_endpoint,
            "recent": recent,
        }


# Singleton shared by middleware + dashboard.
default_logger = CallLogger()
